src/canvasObject.js

```diff
@@ -1,159 +1,152 @@
 'use strict';
 
 require('openseadragon');
+var ImageResource = require('./ImageResource');
 
-var CanvasObject = function(config, dispatcher) {
-  this.fullyOpened = config.fullyOpened || false;
-  this.visible = config.visible || true; // todo: this is not used yet. Do we need it?
+var CanvasObject = function(config) {
+  var self = this;
   this.clipRegion = config.clipRegion;
   this.opacity = config.opacity || 1;
-  this.x = config.x || 0;
-  this.y = config.y || 0;
-  this.placeholder = config.placeholder || { type: 'image', url: './example-thumbnail.png' };
   this.index = config.index;
 
   this.id = config.canvas['@id'];
-  this.height = config.canvas.height;
-  this.width = config.canvas.width;
-  this.images = config.canvas.images;
+
+  this.bounds = {
+    x : config.x || 0,
+    y : config.y || 0,
+    height : config.canvas.height,
+    width : config.canvas.width
+  };
+  this.thumbUrl = config.canvas.thumbnail;
+
+  // todo: Move this logic into an ImageResourceFactory.
+  this._getThumbService = function(width) {
+    var image = config.canvas.images[0];
+    if(image.resource.service) {
+      return image.resource.service['@id'] + '/full/' + Math.ceil(width * 2) + ',/0/default.jpg';
+    } else {
+      return image.resource['@id'];
+    }
+  };
+
   this.label = config.canvas.label;
-  this.tileSourceUrl = config.canvas.images[0].resource.service['@id'] + '/info.json';
   this.viewingHint = config.canvas.viewingHint;
 
-  this.dispatcher = dispatcher;
+  this.dispatcher = config.dispatcher;
+  this.viewer = config.viewer;
+  this.images = [];
+  // details and alternates possibly go here; disambiguate between them.
+  if(config.canvas.images) {
+    this.images = config.canvas.images.map(function(image) {
+
+      // todo: Move this logic into an ImageResourceFactory.
+      var _getImageTilesource = function(image) {
+        if(image.resource.service) {
+          return image.resource.service['@id'] + '/info.json';
+        } else {
+          return image.resource['@id'];
+        }
+      };
+
+      return new ImageResource({
+        tileSource: _getImageTilesource(image),
```

Comment:
It might be worth doing some logic so if the canvas only has a single image (no service) that we don't have both a thumbnail and "real" image, since they're both pointing at the same image file.